Add error reasons to Gatling Highcharts reporter

dancmeyers opened the pull request:
Depends on: https://github.com/mhjort/clojider-gatling-highcharts-reporter/pull/2

This uses the modifications to the Gatling Highcharts reporter
(mhjort/clojider-gatling-highcharts-reporter#2)
to record a breakdown of the types of failure that occur. User-defined
request functions simply need to throw the exception for this to be
triggered.

Given that some libraries (e.g. `httpkit`) return a clojure
`ExceptionInfo` in a failed response it is also useful to check the
response for a directly returned exception object, rather than requiring
it to be thrown.

dancmeyers commented on src/clj_gatling/simulation.clj:
```diff
@@ -16,16 +16,20 @@
-                           {:result (first result) :end-time (now) :context (second result)}
-                           {:result result :end-time (now) :context ctx}))]
+  (let [parse-result   (fn [result]
+                         (if (instance? Exception result)
```
A lot of `Throwable`s we probably don't want to catch, but `AssertionError` is a `Throwable`, not an `Exception`, and people may wish to use `(assert ...` in their steps. What do you think?

mhjort commented:
Thanks! I will take a look. I guess this will solve the issue: https://github.com/mhjort/clj-gatling/issues/62

dancmeyers commented:
Ooh, I hadn't even spotted that one. Yes, it should!

If the same timeout is passed to the httpkit client used in steps and to `clj-gatling` for the timeout channel it's a bit toss-a-coin which actually triggers first, but that just means you might have 2 different types of error reported, both of which are 'actually' timeouts, so it's not a big deal. Regular Gatling has the same issue.

mhjort commented:
Have you tried this with raw reporter: https://github.com/mhjort/clj-gatling#reporters . For some reason in raw report all exceptions are `nil`s.  I have to investigate that more.

dancmeyers commented:
I hadn't noticed that, but with a bit of investigation I can tell you _why_, I'm just not sure of the solution yet.

From my own testing the in-memory raw reporter is fine, this only affects the file-based raw reporter.

You're writing lines of valid EDN to a file, but the file itself is not a vector or similar, so you can't read the whole thing as EDN all at once, instead you have to parse it line-by-line.

`pr-str` writes `Exception` objects as multiline blocks starting with `#error`, which can't actually be parsed by `read-string`. So the raw reporter as-is is writing out entries that span across multiple lines and then attempting to read them line-by-line (problem 1), and also is erroring when it hits `#error` as there's no reader function for that tag (problem 2).

I can do a relatively small change for the file reporter, which is similar to the highcharts one and stringifies the exception but drops the stacktrace etc. That's the easiest and probably the best solution, without going down the route of attempting to write a reader function for `#error` tags, and tell the difference between 'newline between records' and 'newline in a stacktrace',  but as we're losing the stacktrace I don't know if it could still be considered 'raw'.

Also in my own testing the `raw.log` file never got populated, only the `batch-x-y.log` ones, but as that seems to also happen in `master` I've not looked too far into it.

dancmeyers commented on src/clj_gatling/simulation.clj:
```diff
@@ -16,17 +16,23 @@
         (catch Exception e
+          {:result false :end-time (now) :context ctx :exception e})
+        (catch AssertionError e
           {:result false :end-time (now) :context ctx :exception e})))))
```
We catch `AssertionError` so that users can easily make use of `(assert ...)` in their steps, for validation, but `AssertionError` derives directly from `Throwable`, not `Exception`, and we don't want to catch all `Throwable`s, so we don't catch more critical issues.

mhjort replied:
I think this is a non backwards compatible change. I suggest that this change is not part of this PR. This can be an separate issue (e.g. catch also AssertionErrors) where we can have a discussion about this.

dancmeyers replied:
Hmm, OK. I'm not sure why you'd _not_ want to catch `AssertionError`. Previously it would have caused the entire test run to bomb out, so my feeling is that people can't be using `assert`. Or, if they are, are catching it and casting it to a different exception type that is handled before throwing back to `clj-gatling`.

So I guess this is _technically_ a non-backwards-compatible change, I just can't see how anyone could have been usefully using `assert` the old way. Can pull it out to a separate PR though.

dancmeyers replied:
No need to apologise! It's your project/repo ;) And if it makes it clearer for others what is going on, in case they want 0.17.1 and not 0.17.2, then great!

dancmeyers commented on src/clj_gatling/reporters/raw_reporter.clj:
```diff
@@ -19,17 +19,21 @@
+            cleaned-line (update line :requests (fn [reqs]
+                                                  (map #(if (:exception %)
+                                                          (update % :exception str)
+                                                          %) reqs)))]
```
We clean any returned exceptions in the file-based raw reporter by stringifying them. This loses the stacktrace, but allows the log to be parsed correctly. Arguably we already know the source because it's associated with a specific step, and each step is probably quite a small number of lines of code.

mhjort commented:
I had finally time to look this thoroughly. There is a bug with file based raw reporter when running with small simulations (less than 20000 requests). Thanks for noticing. I filed an issue https://github.com/mhjort/clj-gatling/issues/80. It should be easy to fix but didn't have time to fix it now.

I think exceptions add lot of value to reports. However, there are few things to consider:
1) File based and memory based raw reporter should give exactly the same end results. The idea is that you can use memory reporter and then switch to file reporter if you run out of memory.
2) Full exception stack traces can be very verbose. They are already logged fully to `:error-file`.
3) Your point of catching also `AssertionError` is valid. I think it is not backwards compatible change though.

I suggest that we "stringify" exceptions from intermediate results so that reporter implementations never see them.  This way raw reporters work without any modifications. This could be done after logging exception to stack trace and before writing to results channel in line 117. Something like this could work:

```
 (>! result-channel (->> (update result :exception str)
                                    (conj results))))
```

What do you think?

dancmeyers commented:
- Added `Throwable` 'cleaning' to `simulation-util`, plus tests, and used it in `simulation`
- Got rid of the cleaning in the raw reporter, as that now happens in `simulation`.
- Reverted catching of `AssertionError`.

N.B. this now fails to build on CI, as I realised it was still referring to the highcharts bundle 0.3.1 in `deps.edn`, so I updated that but of course that version doesn't yet exist on Clojars.

mhjort commented:
Everything works now as expected. Thanks for the PR. I will merge this now and deploy new version.
